File: tp_3/Correction_chatgpt/welsh_powell.c
#include <stdio.h>
#define MAX_VERTICES 100

struct Graph {
    int vertices;
    int adjMatrix[MAX_VERTICES][MAX_VERTICES];
};

void addEdge(struct Graph* graph, int u, int v) {
    graph->adjMatrix[u][v] = 1;
    graph->adjMatrix[v][u] = 1;
}

void welshPowell(struct Graph* graph) {
    int degree[MAX_VERTICES] = {0};
    int i, j, current_color;

    for (i = 0; i < graph->vertices; i++) {
        for (j = 0; j < graph->vertices; j++) {
            if (graph->adjMatrix[i][j] == 1)
                degree[i]++;
        }
    }

    int sorted_vertices[MAX_VERTICES];

    for (i = 0; i < graph->vertices; i++)
        sorted_vertices[i] = i;

    for (i = 0; i < graph->vertices - 1; i++) {
        for (j = 0; j < graph->vertices - i - 1; j++) {
            if (degree[sorted_vertices[j]] < degree[sorted_vertices[j + 1]]) {
                int temp = sorted_vertices[j];
                sorted_vertices[j] = sorted_vertices[j + 1];
                sorted_vertices[j + 1] = temp;
            }
        }
    }

    int color[MAX_VERTICES];
    
    for (i = 0; i < graph->vertices; i++)
        color[i] = -1;

    current_color = 0;

    for (i = 0; i < graph->vertices; i++) {
        int vertex = sorted_vertices[i];
        
        if (color[vertex] == -1) {

            color[vertex] = current_color;
            
            for (j = i + 1; j < graph->vertices; j++) {
                int v = sorted_vertices[j];
                
                if (color[v] == -1) {
                    if (graph->adjMatrix[vertex][v] == 0) {
                        int can_color = 1;
                        int k = 0;
                        while (k < graph->vertices && can_color) {
                            if (graph->adjMatrix[v][k] == 1 && color[k] == current_color) {
                                can_color = 0;
                            }
                            k++;
                        }
                        if (can_color) {
                            color[v] = current_color;
                        }
                    }
                }
            }
            current_color++;
        }
    }

    int num_colors = current_color;

    printf("Graph colored using %d colors:\n", num_colors);
    for (i = 0; i < graph->vertices; i++) {
        printf("Vertex %d: Color %d\n", i, color[i]);
    }
}

int main() {
    struct Graph graph;
    graph.vertices = 5;

    addEdge(&graph, 0, 1);
    addEdge(&graph, 0, 2);
    addEdge(&graph, 1, 2);
    addEdge(&graph, 1, 3);
    addEdge(&graph, 2, 3);
    addEdge(&graph, 3, 4);
    addEdge(&graph, 1, 4);

    welshPowell(&graph);

    return 0;
}

File: tp_3/Correction_chatgpt/Welsh_powell.py
import networkx as nx
import matplotlib.pyplot as plt


class GraphBacktracking:
    def __init__(self, vertices):
        self.vertices = vertices
        self.adjMatrix = [[0] * vertices for _ in range(vertices)]

    def add_edge(self, u, v):
        self.adjMatrix[u][v] = 1
        self.adjMatrix[v][u] = 1

    def is_safe(self, v, color, c):
        for i in range(self.vertices):
            if self.adjMatrix[v][i] == 1 and color[i] == c:
                return False
        return True

    def graph_coloring_util(self, m, color, v):
        if v == self.vertices:
            return True

        for c in range(1, m+1):
            if self.is_safe(v, color, c):
                color[v] = c
                if self.graph_coloring_util(m, color, v+1):
                    return True
                color[v] = 0

    def graph_coloring(self, m):
        color = [0] * self.vertices
        if not self.graph_coloring_util(m, color, 0):
            return False

        print("Graph can be colored using {} colors:".format(m))
        for i in range(self.vertices):
            print("Vertex {}: Color {}".format(i, color[i]))

        # Visualisation
        G_viz = nx.Graph()
        for i in range(self.vertices):
            for j in range(i+1, self.vertices):
                if self.adjMatrix[i][j] == 1:
                    G_viz.add_edge(i, j)
        
        color_map = ['red', 'blue', 'green', 'yellow', 'orange', 'purple', 'pink', 'brown']
        node_colors = [color_map[color[i] % len(color_map)] for i in range(self.vertices)]
        
        plt.figure(figsize=(8, 6))
        nx.draw(G_viz, with_labels=True, node_color=node_colors, node_size=800, font_size=14, font_weight='bold', edge_color='gray', width=2)
        plt.title(f"Backtracking : {m} couleurs", fontsize=14)
        plt.show()

        return True


class GraphWelshPowell:
    def __init__(self, vertices):
        self.vertices = vertices
        self.adjMatrix = [[0] * vertices for _ in range(vertices)]

    def add_edge(self, u, v):
        self.adjMatrix[u][v] = 1
        self.adjMatrix[v][u] = 1

    def welsh_powell(self):

        degree = []
        for row in self.adjMatrix:
            degree.append(sum(row))
        
        sorted_vertices = sorted(range(self.vertices), key=lambda x: degree[x], reverse=True)

        color = [-1] * self.vertices
        current_color = 0

        for i in range(self.vertices):
            vertex = sorted_vertices[i]
            
            if color[vertex] == -1:
                color[vertex] = current_color
                
                for j in range(i + 1, self.vertices):
                    v = sorted_vertices[j]
                    
                    if color[v] == -1:
                        if self.adjMatrix[vertex][v] == 0:
                            can_color = True
                            k = 0
                            while k < self.vertices and can_color:
                                if self.adjMatrix[v][k] == 1 and color[k] == current_color:
                                    can_color = False
                                k += 1
                            
                            if can_color:
                                color[v] = current_color
                
                current_color += 1

        num_colors = current_color

        print("Graph colored using {} colors:".format(num_colors))
        for i in range(self.vertices):
            print("Vertex {}: Color {}".format(i, color[i]))

        G_viz = nx.Graph()
        for i in range(self.vertices):
            for j in range(i+1, self.vertices):
                if self.adjMatrix[i][j] == 1:
                    G_viz.add_edge(i, j)
        
        color_map = ['red', 'blue', 'green', 'yellow', 'orange', 'purple', 'pink', 'brown']
        node_colors = []
        for i in range(self.vertices):
            node_colors.append(color_map[color[i] % len(color_map)])
        
        plt.figure(figsize=(8, 6))
        nx.draw(G_viz, with_labels=True, node_color=node_colors, node_size=800, font_size=14, font_weight='bold', edge_color='gray', width=2)
        plt.title("Welsh-Powell : {} couleurs".format(num_colors), fontsize=14)
        plt.show()

        return num_colors


if __name__ == "__main__":
    G = nx.Graph()

    G.add_nodes_from([1, 2, 3, 4, 5])

    G.add_edges_from([(1, 2), (2, 3), (3, 4), (4, 5), (5, 1)])

    nx.draw(G, with_labels=True, node_color='lightblue', edge_color='gray')

    plt.show()


    g = GraphBacktracking(5)

    g.add_edge(0, 1)
    g.add_edge(0, 2)
    g.add_edge(1, 2)
    g.add_edge(1, 3)
    g.add_edge(2, 3)
    g.add_edge(3, 4)
    g.add_edge(1, 4)

    m = 4 
    g.graph_coloring(m)

    g = GraphWelshPowell(5)

    g.add_edge(0, 1)
    g.add_edge(0, 2)
    g.add_edge(1, 2)
    g.add_edge(1, 3)
    g.add_edge(2, 3)
    g.add_edge(3, 4)
    g.add_edge(1, 4)

    g.welsh_powell()
